aeron-system-tests/src/test/java/io/aeron/cluster/ClusterTest.java

```diff
@@ -1632,6 +1641,384 @@ public void shouldRecoverWhenFollowersIsMultipleTermsBehindFromEmptyLogAndPartia
         cluster.assertRecordingLogsEqual();
     }
 
+    @Test
+    @InterruptAfter(10)
+    void shouldRejectTakeSnapshotRequestWithAnAuthorisationError()
+    {
+        cluster = aCluster().withStaticNodes(3).start();
+        systemTestWatcher.cluster(cluster);
+
+        final TestNode leader = cluster.awaitLeader();
+        final List<TestNode> followers = cluster.followers();
+
+        final long requestCorrelationId = System.nanoTime();
+        final MutableBoolean responseReceived = new MutableBoolean();
+        cluster.egressListener(new EgressListener()
+        {
+            public void onMessage(
+                final long clusterSessionId,
+                final long timestamp,
+                final DirectBuffer buffer,
+                final int offset,
+                final int length,
+                final Header header)
+            {
+            }
+
+            public void onAdminResponse(
+                final long clusterSessionId,
+                final long correlationId,
+                final AdminRequestType requestType,
+                final AdminResponseCode code,
+                final String message,
+                final DirectBuffer payload,
+                final int payloadOffset,
+                final int payloadLength)
+            {
+                responseReceived.set(true);
+                assertEquals(requestCorrelationId, correlationId);
+                assertEquals(AdminRequestType.SNAPSHOT, requestType);
+                assertEquals(AdminResponseCode.UNAUTHORISED_ACCESS, code);
+                assertEquals("Execution of the " + requestType + " request was not authorised", message);
+                assertNotNull(payload);
+                final int minPayloadOffset =
+                    MessageHeaderEncoder.ENCODED_LENGTH +
+                    AdminResponseEncoder.BLOCK_LENGTH +
+                    AdminResponseEncoder.messageHeaderLength() +
+                    message.length() +
+                    AdminResponseEncoder.payloadHeaderLength();
+                assertTrue(payloadOffset > minPayloadOffset);
+                assertEquals(0, payloadLength);
+            }
+        });
+
+        final AeronCluster client = cluster.connectClient();
+        while (!client.sendAdminRequestToTakeASnapshot(requestCorrelationId))
+        {
+            Tests.yield();
+        }
+
+        while (!responseReceived.get())
+        {
+            client.pollEgress();
+            Tests.yield();
+        }
+
+        assertEquals(0, cluster.getSnapshotCount(leader));
```

Comment:
This sort of validation can be tricky as your are checking to see if something doesn't happen.  The test should pause or retry for a period of time during which it would of been able to take the snapshot if the code was behaving incorrectly - perhaps a second or two.  E.g. I deliberately changed the code to send back an error response, but still carry out the snapshot and this test passed.

Reply:
Good point. I'll update the test.